Mathlib/Data/ValuedCSP.lean

```diff
@@ -0,0 +1,24 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+
+/-!
+
+# General Valued Constraint Satisfaction Problems
+
+-/
+
+/-- Type of functions `α^n → β` for a general arity `n : ℕ` -/
+def ofArity (α β : Type _) : ℕ → Type
+| 0   => β
+| n+1 => α → ofArity α β n
+
+/-- A template for a valued CSP problem with costs in `C` which is usually `ℚ≥0` -/
+structure ValuedCspTemplate (C : Type) [LinearOrderedAddCommMonoid C] where
+  /-- Domain of "labels" -/
+  D : Type
+  /-- Cost functions `D^k → C` for any `k` -/
+  F : Set (Σ (k : ℕ), ofArity D C k)
```

Comment:
I don't think this PR is meaningfully reviewable in its current form; a definition with no API isn't of any use, and you can't tell if a definition is the right one until you start defining/proving things in terms of it.

Reply:
Thanks anyways for having a look!
Let's decide the `ofArity` vs `Arity` thing and then I'll make some API.

Reply:
I strongly suspect the answer is going to be "you don't want `Arity` here anyway"...

Reply:
What is your suspicion based on, please?

Reply:
Arity removed.